File: nix/packages/rice/src/actions/flake.rs
//! Flake update workflow logic.

use anyhow::Result;

use crate::actions::core::{info, ok, run};

/// Update one flake input or all inputs and commit lockfile changes when needed.
pub(crate) fn flake_update(name: &str) -> Result<()> {
    if name == "all" {
        info("Updating all flake inputs...");
        run(
            vec![
                "nix".to_string(),
                "flake".to_string(),
                "update".to_string(),
                "--commit-lock-file".to_string(),
                "--option".to_string(),
                "commit-lockfile-summary".to_string(),
                "chore(flake): update lockfile".to_string(),
            ],
            false,
        )?;
        ok("Flake update complete");
        return Ok(());
    }

    info(&format!("Updating flake input: {name}..."));
    run(
        vec![
            "nix".to_string(),
            "flake".to_string(),
            "update".to_string(),
            name.to_string(),
        ],
        false,
    )?;
    run(
        vec![
            "git".to_string(),
            "add".to_string(),
            "flake.lock".to_string(),
        ],
        false,
    )?;
    run(
        vec![
            "git".to_string(),
            "commit".to_string(),
            "-m".to_string(),
            format!("chore(flake): update input ({name})"),
        ],
        false,
    )?;
    ok("Flake update complete");
    Ok(())
}

File: nix/packages/rice/src/core/actions/flake.ts
import { Effect } from "effect";

import { liveActionRuntime, run, type ActionRuntime } from "./core.ts";

export const flakeUpdate = (
  name: string,
  runtime: ActionRuntime = liveActionRuntime,
): Effect.Effect<void, Error> =>
  Effect.gen(function* () {
    if (name === "all") {
      runtime.console.info("Updating all flake inputs...");
      yield* run(
        [
          "nix",
          "flake",
          "update",
          "--commit-lock-file",
          "--option",
          "commit-lockfile-summary",
          "chore(flake): update lockfile",
        ],
        false,
        runtime,
      );
      runtime.console.ok("Flake update complete");
      return;
    }

    runtime.console.info(`Updating flake input: ${name}...`);
    yield* run(["nix", "flake", "update", name], false, runtime);
    yield* run(["git", "add", "flake.lock"], false, runtime);
    yield* run(["git", "commit", "-m", `chore(flake): update input (${name})`], false, runtime);
    runtime.console.ok("Flake update complete");
  });
